File: ts-src/hooks/runner.ts
import { debug } from "../debug.ts";
import { matchesHook } from "./matcher.ts";
import type { HookContext, HookEvent, HookResult, ResolvedHookDefinition, ResolvedHooksConfig } from "./types.ts";

export class HooksRunner {
	private config: ResolvedHooksConfig;
	private mode: "interactive" | "headless";
	private baseContext: { sessionId: string; project: string; model: string };

	constructor(
		config: ResolvedHooksConfig,
		mode: "interactive" | "headless",
		baseContext: { sessionId: string; project: string; model: string },
	) {
		this.config = config;
		this.mode = mode;
		this.baseContext = baseContext;
	}

	async run(event: HookEvent, context: Partial<HookContext>): Promise<HookResult[]> {
		const hooks = this.config[event];
		if (!hooks || hooks.length === 0) return [];

		const fullContext: HookContext = {
			sessionId: this.baseContext.sessionId,
			project: this.baseContext.project,
			model: this.baseContext.model,
			event,
			...context,
		};

		// Filter by mode
		const modeFiltered = hooks.filter((h) => h.mode === this.mode || h.mode === "both");

		// Filter by matchers
		const matched = modeFiltered.filter((h) => matchesHook(h, fullContext));

		// Split sync and async
		const syncHooks = matched.filter((h) => !h.async);
		const asyncHooks = matched.filter((h) => h.async);

		// Fire async hooks (fire-and-forget)
		for (const hook of asyncHooks) {
			this.executeAsync(hook, fullContext);
		}

		// Execute sync hooks sequentially
		const results: HookResult[] = [];
		for (const hook of syncHooks) {
			const result = await this.executeSync(hook, fullContext);
			results.push(result);
		}

		return results;
	}

	private async executeSync(hook: ResolvedHookDefinition, context: HookContext): Promise<HookResult> {
		const env = this.buildEnv(context);
		const stdinData = this.buildStdin(context);

		try {
			const proc = Bun.spawn(["sh", "-c", hook.command], {
				env: { ...process.env, ...env },
				stdin: "pipe",
				stdout: "pipe",
				stderr: "pipe",
			});

			// Write stdin data and close
			if (stdinData) {
				proc.stdin.write(stdinData);
			}
			proc.stdin.end();

			// Race between process completion and timeout
			const timeoutPromise = new Promise<"timeout">((resolve) => {
				setTimeout(() => resolve("timeout"), hook.timeout);
			});

			const exitPromise = proc.exited.then((code) => ({ code }));
			const raceResult = await Promise.race([exitPromise, timeoutPromise]);

			if (raceResult === "timeout") {
				proc.kill();
				return { blocked: false, timedOut: true };
			}

			const { code } = raceResult;
			const stdout = await new Response(proc.stdout).text();
			const stderr = await new Response(proc.stderr).text();

			if (code !== 0) {
				return {
					blocked: true,
					reason: stderr.trim() || `Hook exited with code ${code}`,
					timedOut: false,
				};
			}

			return {
				blocked: false,
				feedback: stdout.trim() || undefined,
				timedOut: false,
			};
		} catch (err) {
			debug("hooks", `Hook execution error: ${err}`);
			return { blocked: false, timedOut: false };
		}
	}

	private executeAsync(hook: ResolvedHookDefinition, context: HookContext): void {
		const env = this.buildEnv(context);
		const stdinData = this.buildStdin(context);

		try {
			const proc = Bun.spawn(["sh", "-c", hook.command], {
				env: { ...process.env, ...env },
				stdin: "pipe",
				stdout: "ignore",
				stderr: "pipe",
			});

			if (stdinData) {
				proc.stdin.write(stdinData);
			}
			proc.stdin.end();

			// Set up timeout to kill the process
			const timer = setTimeout(() => {
				proc.kill();
			}, hook.timeout);

			// Log warnings on failure (but don't block)
			proc.exited.then((code) => {
				clearTimeout(timer);
				if (code !== 0) {
					new Response(proc.stderr).text().then((stderr) => {
						debug("hooks", `Async hook warning (exit ${code}): ${stderr.trim()}`);
					});
				}
			});
		} catch (err) {
			debug("hooks", `Async hook spawn error: ${err}`);
		}
	}

	private buildEnv(context: HookContext): Record<string, string> {
		const env: Record<string, string> = {
			HEDDLE_HOOK_EVENT: context.event,
			HEDDLE_HOOK_SESSION_ID: context.sessionId,
			HEDDLE_HOOK_PROJECT: context.project,
			HEDDLE_HOOK_MODEL: context.model,
		};
		if (context.toolName) {
			env.HEDDLE_HOOK_TOOL_NAME = context.toolName;
		}
		return env;
	}

	private buildStdin(context: HookContext): string | null {
		const data: Record<string, string | undefined> = {};
		let hasData = false;

		if (context.toolArgs !== undefined) {
			data.tool_args = context.toolArgs;
			hasData = true;
		}
		if (context.toolResult !== undefined) {
			data.tool_result = context.toolResult;
			hasData = true;
		}
		if (context.userInput !== undefined) {
			data.user_input = context.userInput;
			hasData = true;
		}

		return hasData ? JSON.stringify(data) : null;
	}
}

File: src/hooks/runner.rs
//! HooksRunner — spawns hook commands and collects their results.
//! Mirrors `ts-src/hooks/runner.ts`.

use std::collections::HashMap;
use std::process::Stdio;
use std::time::Duration;

use tokio::io::AsyncWriteExt;
use tokio::process::Command;
use tokio::time::timeout;

use super::matcher::matches_hook;
use super::types::{
    HookContext, HookEvent, HookMode, HookResult, ResolvedHookDefinition, ResolvedHooksConfig,
};
use crate::debug::debug;

pub struct HooksRunner {
    pub config: ResolvedHooksConfig,
    pub mode: HookMode,
    pub session_id: String,
    pub project: String,
    pub model: String,
}

impl HooksRunner {
    pub fn new(
        config: ResolvedHooksConfig,
        mode: HookMode,
        session_id: String,
        project: String,
        model: String,
    ) -> Self {
        Self {
            config,
            mode,
            session_id,
            project,
            model,
        }
    }

    pub async fn run(&self, event: HookEvent, mut context: HookContext) -> Vec<HookResult> {
        let hooks = match self.config.get(&event) {
            Some(h) if !h.is_empty() => h,
            _ => return Vec::new(),
        };

        context.session_id = self.session_id.clone();
        context.project = self.project.clone();
        context.model = self.model.clone();
        context.event = event.as_str().to_string();

        // Filter by mode
        let mode_matched: Vec<&ResolvedHookDefinition> = hooks
            .iter()
            .filter(|h| matches!(h.mode, HookMode::Both) || h.mode == self.mode)
            .collect();

        // Filter by matchers
        let matched: Vec<&ResolvedHookDefinition> = mode_matched
            .into_iter()
            .filter(|h| matches_hook(h, &context))
            .collect();

        // Fire async hooks (fire-and-forget)
        for hook in matched.iter().filter(|h| h.r#async) {
            let hook = (*hook).clone();
            let ctx = context.clone();
            tokio::spawn(async move {
                execute_async(&hook, &ctx).await;
            });
        }

        // Sync hooks sequentially
        let mut results = Vec::new();
        for hook in matched.iter().filter(|h| !h.r#async) {
            let result = execute_sync(hook, &context).await;
            results.push(result);
        }
        results
    }
}

fn build_env(context: &HookContext) -> HashMap<String, String> {
    let mut env = HashMap::new();
    env.insert("HEDDLE_HOOK_EVENT".to_string(), context.event.clone());
    env.insert(
        "HEDDLE_HOOK_SESSION_ID".to_string(),
        context.session_id.clone(),
    );
    env.insert("HEDDLE_HOOK_PROJECT".to_string(), context.project.clone());
    env.insert("HEDDLE_HOOK_MODEL".to_string(), context.model.clone());
    if let Some(t) = &context.tool_name {
        env.insert("HEDDLE_HOOK_TOOL_NAME".to_string(), t.clone());
    }
    env
}

fn build_stdin(context: &HookContext) -> Option<String> {
    let mut map = serde_json::Map::new();
    if let Some(v) = &context.tool_args {
        map.insert(
            "tool_args".to_string(),
            serde_json::Value::String(v.clone()),
        );
    }
    if let Some(v) = &context.tool_result {
        map.insert(
            "tool_result".to_string(),
            serde_json::Value::String(v.clone()),
        );
    }
    if let Some(v) = &context.user_input {
        map.insert(
            "user_input".to_string(),
            serde_json::Value::String(v.clone()),
        );
    }
    if map.is_empty() {
        None
    } else {
        Some(serde_json::Value::Object(map).to_string())
    }
}

async fn execute_sync(hook: &ResolvedHookDefinition, ctx: &HookContext) -> HookResult {
    let env = build_env(ctx);
    let stdin_data = build_stdin(ctx);

    let mut cmd = Command::new("sh");
    cmd.args(["-c", &hook.command])
        .envs(env.iter())
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    let mut child = match cmd.spawn() {
        Ok(c) => c,
        Err(e) => {
            debug("hooks", &format!("Hook spawn error: {e}"));
            return HookResult::default();
        }
    };

    if let Some(data) = stdin_data {
        if let Some(stdin) = child.stdin.as_mut() {
            let _ = stdin.write_all(data.as_bytes()).await;
        }
    }
    if let Some(stdin) = child.stdin.take() {
        drop(stdin);
    }

    let timeout_dur = Duration::from_millis(hook.timeout);
    let waited = timeout(timeout_dur, child.wait_with_output()).await;

    let output = match waited {
        Ok(Ok(out)) => out,
        Ok(Err(e)) => {
            debug("hooks", &format!("Hook wait error: {e}"));
            return HookResult::default();
        }
        Err(_) => {
            // Timed out — kill is implicit via tokio when child is dropped, but
            // we won't have a kill handle here. The contract says "timed_out".
            return HookResult {
                timed_out: true,
                ..Default::default()
            };
        }
    };

    let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
    let code = output.status.code().unwrap_or(-1);

    if code != 0 {
        return HookResult {
            blocked: true,
            reason: Some(if stderr.is_empty() {
                format!("Hook exited with code {code}")
            } else {
                stderr
            }),
            ..Default::default()
        };
    }
    HookResult {
        feedback: if stdout.is_empty() {
            None
        } else {
            Some(stdout)
        },
        ..Default::default()
    }
}

async fn execute_async(hook: &ResolvedHookDefinition, ctx: &HookContext) {
    let env = build_env(ctx);
    let stdin_data = build_stdin(ctx);

    let mut cmd = Command::new("sh");
    cmd.args(["-c", &hook.command])
        .envs(env.iter())
        .stdin(Stdio::piped())
        .stdout(Stdio::null())
        .stderr(Stdio::piped());

    let mut child = match cmd.spawn() {
        Ok(c) => c,
        Err(e) => {
            debug("hooks", &format!("Async hook spawn error: {e}"));
            return;
        }
    };

    if let Some(data) = stdin_data {
        if let Some(stdin) = child.stdin.as_mut() {
            let _ = stdin.write_all(data.as_bytes()).await;
        }
    }
    drop(child.stdin.take());

    let timeout_dur = Duration::from_millis(hook.timeout);
    let _ = timeout(timeout_dur, child.wait_with_output()).await;
}
